src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -233,10 +236,391 @@ private SyntaxList<RazorSyntaxNode> ParseTagBody(string tagName, out bool seenEn
+                                if (!EndOfFile)
+                                {
+                                    Assert(openQuoteKind);
+                                    AcceptAndMoveNext();
+                                }
+                                break;
+                            case SyntaxKind.OpenAngle: // Another "<" means this tag is invalid.
+                            case SyntaxKind.ForwardSlash: // Empty tag
+                            case SyntaxKind.CloseAngle: // End of tag
+                                miscAttributeContentBuilder.Add(OutputAsMarkupLiteral());
+                                if (miscAttributeContentBuilder.Count > 0)
+                                {
+                                    miscAttributeContent = SyntaxFactory.MarkupMiscAttributeContent(miscAttributeContentBuilder.ToList());
+                                    builder.Add(miscAttributeContent);
+                                }
+                                return;
+                            default:
+                                AcceptAndMoveNext();
+                                break;
+                        }
+                    }
+                }
+
+                miscAttributeContentBuilder.Add(OutputAsMarkupLiteral());
+                if (miscAttributeContentBuilder.Count > 0)
+                {
+                    miscAttributeContent = SyntaxFactory.MarkupMiscAttributeContent(miscAttributeContentBuilder.ToList());
+                    builder.Add(miscAttributeContent);
+                }
+            }
+        }
+
+        private bool IsTagRecoveryStopPoint(SyntaxToken token)
+        {
+            return token.Kind == SyntaxKind.CloseAngle ||
+                   token.Kind == SyntaxKind.ForwardSlash ||
+                   token.Kind == SyntaxKind.OpenAngle ||
+                   token.Kind == SyntaxKind.SingleQuote ||
+                   token.Kind == SyntaxKind.DoubleQuote;
+        }
+
+        private void BeforeAttribute(in SyntaxListBuilder<RazorSyntaxNode> builder)
+        {
+            // http://dev.w3.org/html5/spec/tokenization.html#before-attribute-name-state
+            // Capture whitespace
+            var whitespace = ReadWhile(token => token.Kind == SyntaxKind.Whitespace || token.Kind == SyntaxKind.NewLine);
+
+            if (At(SyntaxKind.Transition) || At(SyntaxKind.RazorCommentTransition))
+            {
+                // Transition outside of attribute value => Switch to recovery mode
+                Accept(whitespace);
+                ParseMiscAttribute(builder);
+                return;
+            }
+
+            // http://dev.w3.org/html5/spec/tokenization.html#attribute-name-state
+            // Read the 'name' (i.e. read until the '=' or whitespace/newline)
+            var nameTokens = Enumerable.Empty<SyntaxToken>();
+            var whitespaceAfterAttributeName = Enumerable.Empty<SyntaxToken>();
+            if (IsValidAttributeNameToken(CurrentToken))
```

Comment:
I tried it just now but it makes the contract of TryParseAttributeName complicated because of the preceding whitespace and introduces magic. I prefer leaving the transition/comment check outside.

Reply:
I was wrong. Taylor: 1 Ajay: 0 😄